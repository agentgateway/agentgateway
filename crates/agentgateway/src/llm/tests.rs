use std::fs;
use std::path::Path;

use serde::de::DeserializeOwned;

use super::*;

fn test_response<T: DeserializeOwned>(
	test_name: &str,
	xlate: impl Fn(T) -> Result<universal::ChatCompletionResponse, AIError>,
) {
	let test_dir = Path::new("src/llm/tests");

	// Read input JSON
	let input_path = test_dir.join(format!("{test_name}.json"));
	let provider_str = &fs::read_to_string(&input_path).expect("Failed to read input file");
	let provider_raw: Value =
		serde_json::from_str(provider_str).expect("Failed to parse provider json");
	let provider: T = serde_json::from_str(provider_str).expect("Failed to parse provider JSON");

	let openai_response =
		xlate(provider).expect("Failed to translate provider response to OpenAI format");

	insta::with_settings!({
			info => &provider_raw,
			description => input_path.to_string_lossy().to_string(),
			omit_expression => true,
			prepend_module_to_snapshot => false,
			snapshot_path => "tests",
	}, {
			 insta::assert_json_snapshot!(test_name, openai_response, {
			".id" => "[id]",
			".created" => "[date]",
		});
	});
}

fn test_request<T: Serialize>(
	provider_name: &str,
	test_name: &str,
	xlate: impl Fn(universal::ChatCompletionRequest) -> Result<T, AIError>,
) {
	let test_dir = Path::new("src/llm/tests");

	// Read input JSON
	let input_path = test_dir.join(format!("{test_name}.json"));
	let openai_str = &fs::read_to_string(&input_path).expect("Failed to read input file");
	let openai_raw: Value = serde_json::from_str(openai_str).expect("Failed to parse openai json");
	let openai: universal::ChatCompletionRequest =
		serde_json::from_str(openai_str).expect("Failed to parse openai JSON");

	let provider_response =
		xlate(openai).expect("Failed to translate OpenAI format to provider request ");

	insta::with_settings!({
			info => &openai_raw,
			description => format!("{}: {}", provider_name, test_name),
			omit_expression => true,
			prepend_module_to_snapshot => false,
			snapshot_path => "tests",
	}, {
			 insta::assert_json_snapshot!(format!("{}-{}", provider_name, test_name), provider_response, {
			".id" => "[id]",
			".created" => "[date]",
		});
	});
}

#[test]
fn test_bedrock() {
	let response = |i| bedrock::translate_response(i, &strng::new("fake-model"));
	test_response::<bedrock::types::ConverseResponse>("basic_bedrock", response);

	let request = |i| Ok(bedrock::translate_request(i));
	test_request("bedrock", "basic_input", request);
	test_request("bedrock", "full_input", request);
}

#[test]
fn test_anthropic() {
	let response = |i| Ok(anthropic::translate_response(i));
	test_response::<anthropic::types::MessagesResponse>("basic_anthropic", response);

	let request = |i| Ok(anthropic::translate_request(i));
	test_request("anthropic", "basic_input", request);
	test_request("anthropic", "full_input", request);
}
