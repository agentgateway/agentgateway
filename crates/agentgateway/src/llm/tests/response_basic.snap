---
source: crates/agentgateway/src/llm/tests.rs
description: src/llm/tests/response_basic.json
info:
  id: chatcmpl-C3kBOLGARFj2WJrm5MRyuTiK4u6eF
  object: chat.completion
  created: 1755008546
  model: gpt-3.5-turbo-0125
  choices:
    - index: 0
      message:
        role: assistant
        content: "Sorry, I couldn't find the name of the LLM provider. Could you please provide more information or context?"
        refusal: ~
        annotations: []
      logprobs: ~
      finish_reason: stop
  usage:
    prompt_tokens: 17
    completion_tokens: 23
    total_tokens: 40
    prompt_tokens_details:
      cached_tokens: 0
      audio_tokens: 0
    completion_tokens_details:
      reasoning_tokens: 0
      audio_tokens: 0
      accepted_prediction_tokens: 0
      rejected_prediction_tokens: 0
  service_tier: default
  system_fingerprint: ~
---
{
  "id": "[id]",
  "choices": [
    {
      "index": 0,
      "message": {
        "content": "Sorry, I couldn't find the name of the LLM provider. Could you please provide more information or context?",
        "role": "assistant",
        "annotations": []
      },
      "finish_reason": "stop"
    }
  ],
  "created": "[date]",
  "model": "gpt-3.5-turbo-0125",
  "service_tier": "default",
  "object": "chat.completion",
  "usage": {
    "prompt_tokens": 17,
    "completion_tokens": 23,
    "total_tokens": 40,
    "prompt_tokens_details": {
      "audio_tokens": 0,
      "cached_tokens": 0
    },
    "completion_tokens_details": {
      "accepted_prediction_tokens": 0,
      "audio_tokens": 0,
      "reasoning_tokens": 0,
      "rejected_prediction_tokens": 0
    }
  }
}
